tools/fuzzer/main.c

```diff
@@ -0,0 +1,342 @@
+	int i;
+
+	for (i = 0; i < plat->num_mem_regions; i++)
+		qemu_io_free_shm(i);
+
+	for (i = 0; i < plat->num_reg_regions; i++)
+		qemu_io_free_shm(i);
+
+	qemu_io_free();
+}
+
+/* called by platform when it receives IPC message */
+void fuzzer_ipc_msg_rx(struct fuzz *fuzzer)
+{
+	struct sof_ipc_comp_reply r;
+	struct sof_ipc_cmd_hdr hdr;
+	uint32_t cmd;
+
+	/* read mailbox */
+	fuzzer->platform->mailbox_read(fuzzer, 0, &hdr, sizeof(hdr));
+	cmd = hdr.cmd & SOF_GLB_TYPE_MASK;
+
+	/* check message type */
+	switch (cmd) {
+	case SOF_IPC_GLB_REPLY:
+		fprintf(stderr, "error: ipc reply unknown\n");
+		break;
+	case SOF_IPC_FW_READY:
+		fuzzer->platform->fw_ready(fuzzer);
+		fuzzer->boot_complete = 1;
+		break;
+	case SOF_IPC_GLB_COMPOUND:
+	case SOF_IPC_GLB_TPLG_MSG:
+	case SOF_IPC_GLB_PM_MSG:
+	case SOF_IPC_GLB_COMP_MSG:
+	case SOF_IPC_GLB_STREAM_MSG:
+	case SOF_IPC_GLB_TRACE_MSG:
+		fuzzer->platform->mailbox_read(fuzzer, 0, &r, sizeof(r));
+		break;
+	default:
+		fprintf(stderr, "error: unknown DSP message 0x%x\n", cmd);
+		break;
+	}
+}
+
+/* called by platform when it receives IPC message reply */
+void fuzzer_ipc_msg_reply(struct fuzz *fuzzer)
+{
+	int ret;
+
+	ret = fuzzer->platform->get_reply(fuzzer, &fuzzer->msg);
+	if (ret < 0)
+		fprintf(stderr, "error: incorrect DSP reply\n");
+
+	ipc_dump(&fuzzer->msg);
+
+	pthread_mutex_lock(&ipc_mutex);
+	ipccond_var = 1;
+	pthread_cond_signal(&ipc_cond);
+	pthread_mutex_unlock(&ipc_mutex);
+}
+
+/* called by platform when FW crashses */
+void fuzzer_ipc_crash(struct fuzz *fuzzer, unsigned int offset)
+{
+	/* TODO: DSP FW has crashed. dump stack, regs, last IPC, log etc */
+	fprintf(stderr, "DSP FW crash\n");
+	exit(EXIT_FAILURE);
+}
+
+int fuzzer_send_msg(struct fuzz *fuzzer)
+{
+	struct timespec timeout;
+	struct timeval tp;
+	int ret;
+
+	ipc_dump(&fuzzer->msg);
+
+	/* send msg */
+	ret = fuzzer->platform->send_msg(fuzzer, &fuzzer->msg);
+	if (ret < 0) {
+		fprintf(stderr, "error: message tx failed\n");
+		return ret;
+	}
+
+	/* wait for ipc reply */
+	gettimeofday(&tp, NULL);
+	timeout.tv_sec  = tp.tv_sec;
+	timeout.tv_nsec = tp.tv_usec * 1000;
+	timeout.tv_nsec += 300000000; /* 300ms timeout */
+
+	/* first lock the boot wait mutex */
+	pthread_mutex_lock(&ipc_mutex);
+
+	/* now wait for mutex unlock and ipccond_var set by IPC reply */
+	ret = 0;
+	while (!ipccond_var && ret)
+		ret = pthread_cond_timedwait(&ipc_cond, &ipc_mutex, &timeout);
+
+	if (ret == ETIMEDOUT && !ipccond_var) {
+		ret = -EINVAL;
+		fprintf(stderr, "error: IPC timeout\n");
+		ipc_dump_err(&fuzzer->msg);
+		pthread_mutex_unlock(&ipc_mutex);
+		exit(0);
+	}
+
+	/* reset condition for next IPC message */
+	ipccond_var = 0;
+
+	pthread_mutex_unlock(&ipc_mutex);
+
+	/*
+	 * sleep for 5 ms before continuing sending the next message.
+	 * This helps with the condition signaling. Without this,
+	 * the condition seems to always satisfy and
+	 * the fuzzer never waits for a response from the DSP.
+	 */
+	usleep(50000);
```

Comment:
Did you try to remove this? Should be ok now.

Reply:
@kv2019i i still cant remove this. I'm not sure how to get rid of this. Do you mind taking a look?